[JSC] Add fallback FastStringifier for 16-bit strings

Constellation opened the pull request:
#### f5cf4d7dd0874c15e293967d0f368e0cd303d997
<pre>
[JSC] Add fallback FastStringifier for 16-bit strings
<a href="https://bugs.webkit.org/show_bug.cgi?id=260924">https://bugs.webkit.org/show_bug.cgi?id=260924</a>
rdar://114723262

Reviewed by Mark Lam.

This patch enhances FastStringifier to accept 16-bit strings.
We make it FastStringifier&lt;CharType&gt;, and keep it the previous 8-bit strings fast path.
And we fall back to FastStringifier&lt;UChar&gt; only when the following conditions are met.

    1. We failed due to 16-bit string.
    2. The content is filled less than 50% of the buffer.

We also adjust bufferSize from 6000 to 8192. The previous value comes from Speedometer2.1.
The new value comes from Speedometer3. And we also support undefined property skipping in the FastStringifier.

* JSTests/stress/json-undefined-prop-skip.js: Added.
(shouldBe):
(shouldBe.JSON.stringify):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::JSONObject::JSONObject):
(JSC::FastStringifier&lt;CharType&gt;::logOutcome):
(JSC::logOutcomeImpl):
(JSC::FastStringifier&lt;CharType&gt;::usableBufferSize):
(JSC::FastStringifier&lt;CharType&gt;::FastStringifier):
(JSC::FastStringifier&lt;CharType&gt;::haveFailure const):
(JSC::FastStringifier&lt;CharType&gt;::result const):
(JSC::FastStringifier&lt;CharType&gt;::recordFailure):
(JSC::FastStringifier&lt;CharType&gt;::recordBufferFull):
(JSC::FastStringifier&lt;CharType&gt;::hasRemainingCapacity):
(JSC::FastStringifier&lt;CharType&gt;::hasRemainingCapacitySlow):
(JSC::FastStringifier&lt;CharType&gt;::recordFastPropertyEnumerationFailure):
(JSC::FastStringifier&lt;CharType&gt;::firstGetterSetterPropertyName const):
(JSC::FastStringifier&lt;CharType&gt;::mayHaveToJSON const):
(JSC::FastStringifier&lt;CharType&gt;::append):
(JSC::FastStringifier&lt;CharType&gt;::stringify):
(JSC::stringify):
(JSC::FastStringifier::logOutcome): Deleted.
(JSC::FastStringifier::usableBufferSize): Deleted.
(JSC::FastStringifier::FastStringifier): Deleted.
(JSC::FastStringifier::haveFailure const): Deleted.
(JSC::FastStringifier::result const): Deleted.
(JSC::FastStringifier::recordFailure): Deleted.
(JSC::FastStringifier::recordBufferFull): Deleted.
(JSC::FastStringifier::hasRemainingCapacity): Deleted.
(JSC::FastStringifier::hasRemainingCapacitySlow): Deleted.
(JSC::FastStringifier::recordFastPropertyEnumerationFailure): Deleted.
(JSC::FastStringifier::firstGetterSetterPropertyName const): Deleted.
(JSC::FastStringifier::mayHaveToJSON const): Deleted.
(JSC::FastStringifier::append): Deleted.
(JSC::FastStringifier::stringify): Deleted.

Canonical link: <a href="https://commits.webkit.org/267481@main">https://commits.webkit.org/267481@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8525081deee305f56cc208dee8e18abb6f89e47

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/16720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17044 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17493 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/18502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/15667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20276 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17183 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/18502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16919 "Failed to checkout and rebase branch from PR 17247") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/17293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/14467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19289 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/14537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/15151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/14420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/15526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/15317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/19617 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/15950 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/15922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/18279 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/15103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/4276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/19469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/19501 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2064 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/15753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/3/builds/4123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -62,7 +62,6 @@ JSONObject::JSONObject(VM& vm, Structure* structure)
     : JSNonFinalObject(vm, structure)
 {
 }
-
```
Undo this.

Constellation replied:
Done.

MenloDorian commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -979,10 +1005,18 @@ void FastStringifier::append(JSValue value)
+            auto result = std::to_chars(temporary.data(), temporary.data() + maxInt32StringLength, number);
+            ASSERT(result.ec != std::errc::value_too_large);
+            WTF::copyElements(bitwise_cast<uint16_t*>(&m_buffer[m_length]), bitwise_cast<const uint8_t*>(temporary.data()), result.ptr - temporary.data());
+            m_length += result.ptr - temporary.data();
```
nit: precompute the `lengthToCopy` just once so that the code shows that both of these are using the same value?

Constellation replied:
Done.

MenloDorian commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -1015,28 +1057,66 @@ void FastStringifier::append(JSValue value)
+                for (unsigned i = 0; i < stringLength; ++i) {
+                    auto character = characters[i];
+                    if (UNLIKELY(U16_IS_SURROGATE(character))) {
+                        recordFailure("string is surroage"_s);
```
typo: surroage -> surrogate

Constellation replied:
Done.

MenloDorian commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -1080,10 +1160,22 @@ void FastStringifier::append(JSValue value)
                 return false;
             }
+
+            // Right now, we do not support 16-bit name here since name in 16-bit is much rare than 16-bit string.
```
I suggest /much rare/significantly more rare/

Constellation replied:
Changed.

MenloDorian reviewed (approved):
r=me with fixes.

Constellation committed:
[JSC] Add fallback FastStringifier for 16-bit strings
https://bugs.webkit.org/show_bug.cgi?id=260924
rdar://114723262

Reviewed by Mark Lam.

This patch enhances FastStringifier to accept 16-bit strings.
We make it FastStringifier<CharType>, and keep it the previous 8-bit strings fast path.
And we fall back to FastStringifier<UChar> only when the following conditions are met.

    1. We failed due to 16-bit string.
    2. The content is filled less than 50% of the buffer.

We also adjust bufferSize from 6000 to 8192. The previous value comes from Speedometer2.1.
The new value comes from Speedometer3. And we also support undefined property skipping in the FastStringifier.

* JSTests/stress/json-undefined-prop-skip.js: Added.
(shouldBe):
(shouldBe.JSON.stringify):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::JSONObject::JSONObject):
(JSC::FastStringifier<CharType>::logOutcome):
(JSC::logOutcomeImpl):
(JSC::FastStringifier<CharType>::usableBufferSize):
(JSC::FastStringifier<CharType>::FastStringifier):
(JSC::FastStringifier<CharType>::haveFailure const):
(JSC::FastStringifier<CharType>::result const):
(JSC::FastStringifier<CharType>::recordFailure):
(JSC::FastStringifier<CharType>::recordBufferFull):
(JSC::FastStringifier<CharType>::hasRemainingCapacity):
(JSC::FastStringifier<CharType>::hasRemainingCapacitySlow):
(JSC::FastStringifier<CharType>::recordFastPropertyEnumerationFailure):
(JSC::FastStringifier<CharType>::firstGetterSetterPropertyName const):
(JSC::FastStringifier<CharType>::mayHaveToJSON const):
(JSC::FastStringifier<CharType>::append):
(JSC::FastStringifier<CharType>::stringify):
(JSC::stringify):
(JSC::FastStringifier::logOutcome): Deleted.
(JSC::FastStringifier::usableBufferSize): Deleted.
(JSC::FastStringifier::FastStringifier): Deleted.
(JSC::FastStringifier::haveFailure const): Deleted.
(JSC::FastStringifier::result const): Deleted.
(JSC::FastStringifier::recordFailure): Deleted.
(JSC::FastStringifier::recordBufferFull): Deleted.
(JSC::FastStringifier::hasRemainingCapacity): Deleted.
(JSC::FastStringifier::hasRemainingCapacitySlow): Deleted.
(JSC::FastStringifier::recordFastPropertyEnumerationFailure): Deleted.
(JSC::FastStringifier::firstGetterSetterPropertyName const): Deleted.
(JSC::FastStringifier::mayHaveToJSON const): Deleted.
(JSC::FastStringifier::append): Deleted.
(JSC::FastStringifier::stringify): Deleted.

Canonical link: https://commits.webkit.org/267481@main
